From 93e862bd718c63adc0260c1127279dc25347de67 Mon Sep 17 00:00:00 2001
From: Matt Fischer <matt@mattfischer.com>
Date: Tue, 30 Jun 2015 09:50:49 -0600
Subject: [PATCH] Support setting instance_user to an empty string

This deals with the case where you don't want heat setting instance_user
for you. See
http://lists.openstack.org/pipermail/openstack-dev/2015-June/068333.html
for details.

Closes-Bug: #1472053

Change-Id: I9e8be0dd50709d271fc81683770c78380724e405
---
 heat/manifests/init.pp              |  9 +++++++--
 heat/spec/classes/heat_init_spec.rb | 14 +++++++++++++-
 2 files changed, 20 insertions(+), 3 deletions(-)

diff --git a/heat/manifests/init.pp b/heat/manifests/init.pp
index b6a7327..7e5f4a8 100644
--- a/heat/manifests/init.pp
+++ b/heat/manifests/init.pp
@@ -154,7 +154,9 @@
 # [*instance_user*]
 #   (Optional) The default user for new instances. Although heat claims that
 #   this feature is deprecated, it still sets the users to ec2-user if
-#   you leave this unset. This will likely be deprecated in K or L.
+#   you leave this unset. If you want heat to not set instance_user to
+#   ec2-user, you need to set this to an empty string. This feature has been
+#   deprecated for some time and will likely be removed in L or M.
 #
 # [*enable_stack_adopt*]
 #   (Optional) Enable the stack-adopt feature.
@@ -552,7 +554,10 @@ class heat(
   }
 
   # instance_user
-  if $instance_user {
+  # special case for empty string since it's a valid value
+  if $instance_user == '' {
+    heat_config { 'DEFAULT/instance_user': value => ''; }
+  } elsif $instance_user {
     heat_config { 'DEFAULT/instance_user': value => $instance_user; }
   } else {
     heat_config { 'DEFAULT/instance_user': ensure => absent; }
diff --git a/heat/spec/classes/heat_init_spec.rb b/heat/spec/classes/heat_init_spec.rb
index 83f3307..c150f3e 100644
--- a/heat/spec/classes/heat_init_spec.rb
+++ b/heat/spec/classes/heat_init_spec.rb
@@ -457,7 +457,7 @@ describe 'heat' do
     end
   end
 
-  shared_examples_for 'with instance_user set' do
+  shared_examples_for 'with instance_user set to a string' do
     before do
       params.merge!(
         :instance_user => "fred",
@@ -469,6 +469,18 @@ describe 'heat' do
     end
   end
 
+  shared_examples_for 'with instance_user set to an empty string' do
+    before do
+      params.merge!(
+        :instance_user => "",
+      )
+    end
+
+    it 'has instance_user set to an empty string when specified' do
+      is_expected.to contain_heat_config('DEFAULT/instance_user').with_value('')
+    end
+  end
+
   shared_examples_for 'without instance_user set' do
     it 'doesnt have instance_user set by default' do
       is_expected.to contain_heat_config('DEFAULT/instance_user').with_enure('absent')
